prefect_aws/filesystems.py

```diff
@@ -0,0 +1,201 @@
+    aws_credentials: Optional[AwsCredentials]
+    basepath: Optional[Path]
+    endpoint_url: Optional[str]
+
+    @validator("basepath", pre=True)
+    def cast_pathlib(cls, value):
+
+        """
+        If basepath provided, it means we aren't writing to the root directory
+        of the bucket. We need to ensure that it is a valid path. This is called
+        when the S3Bucket block is instantiated.
+        """
+
+        if isinstance(value, Path):
+            return str(value)
+        return value
+
+    def _resolve_path(self, path: str) -> Path:
+
+        """
+        A helper function used in write_path to resolve the basepath and key.
+
+        Args:
+
+            - path: Name of the key, e.g. "file1". Each object in your
+            bucket has a unique key (or key name).
+
+        Example:
+
+        ```python
+            s3_bucket_block = S3Bucket(
+                bucket_name="bucket",
+                aws_credentials=AwsCredentials,
+                basepath="subfolder"
+            )
+            s3_bucket_block.write_path(path="file1", content=data)
+        ```
+
+            _resolve_path("file1") in this example would return "subfolder/file1"
+        """
+
+        path = path or str(uuid4())
+
+        # If basepath provided, it means we won't write to the root dir of
+        # the bucket. So we need to add it on the front of the path.
+        path = str(Path(self.basepath) / path) if self.basepath else path
+
+        return path
+
+    def _get_s3_client(self) -> boto3.client:
+
+        """
+        Authenticate MinIO credentials or AWS credentials and return an S3 client.
+        This is a helper function called by read_path() or write_path().
+        """
+
+        s3_client_kwargs = {}
+
+        if self.minio_credentials:
+
+            aws_secret_access_key = self.minio_credentials.minio_root_password
+            s3_client_kwargs.update(
+                aws_access_key_id=self.minio_credentials.minio_root_user,
+                aws_secret_access_key=aws_secret_access_key.get_secret_value(),
+                endpoint_url=self.endpoint_url,
+            )
+
+        s3_client = boto3.client(service_name="s3", **s3_client_kwargs)
+
+        return s3_client
+
+    async def read_path(self, path: str) -> bytes:
+
+        """
+        Read specified path from S3 and return contents. Provide the entire
+        path to the key in S3.
+
+        Args:
+            - path: Entire path to (and including) the key.
+
+        Example:
+
+        If I have a bucket called "bucket" containing a folder called
+        "subfolder" which contains an object called "file1", I can
+        access its contents like so:
+
+        ```python
+        s3_bucket_block = S3Bucket(
+            bucket_name="bucket",
+            aws_credentials=AwsCredentials,
+            basepath="subfolder"
+        )
+
+        key_contents = s3_bucket_block.read_path(path="subfolder/file1")
```

Comment:
```suggestion
        Example:
            Read "subfolder/file1" contents from an S3 bucket named "bucket":
            ```python
            s3_bucket_block = S3Bucket(
                bucket_name="bucket",
                aws_credentials=AwsCredentials,
                basepath="subfolder"
            )
    
            key_contents = s3_bucket_block.read_path(path="subfolder/file1")
```